pkg/spacebinding/spacebindinglister_test.go

```diff
@@ -0,0 +1,222 @@
+package spacebinding_test
+
+import (
+	"fmt"
+	"testing"
+
+	toolchainv1alpha1 "github.com/codeready-toolchain/api/api/v1alpha1"
+	"github.com/codeready-toolchain/toolchain-common/pkg/spacebinding"
+	"github.com/codeready-toolchain/toolchain-common/pkg/test"
+	. "github.com/codeready-toolchain/toolchain-common/pkg/test/masteruserrecord"
+	spacetest "github.com/codeready-toolchain/toolchain-common/pkg/test/space"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestNewSpaceBindingLister(t *testing.T) {
+
+	t.Run("recursive list for space", func(t *testing.T) {
+		// given
+		mur := NewMasterUserRecord(t, "johnny", TargetCluster(test.MemberClusterName), TierName("deactivate90"))
+		// we have a parent space
+		parentSpace := spacetest.NewSpace(test.HostOperatorNs, "smith",
+			spacetest.WithTierName("advanced"),
+			spacetest.WithSpecTargetCluster(test.MemberClusterName),
+			spacetest.WithLabel(toolchainv1alpha1.SpaceCreatorLabelKey, "johnny"),
+		)
+		// we have 2 subspaces for the above parent space
+		childSpace1 := spacetest.NewSpace(test.HostOperatorNs, "smith-child1",
+			spacetest.WithTierName("advanced"),
+			spacetest.WithSpecTargetCluster(test.MemberClusterName),
+			spacetest.WithSpecParentSpace(parentSpace.GetName()),
+		)
+		childSpace2 := spacetest.NewSpace(test.HostOperatorNs, "smith-child2",
+			spacetest.WithTierName("advanced"),
+			spacetest.WithSpecTargetCluster(test.MemberClusterName),
+			spacetest.WithSpecParentSpace(parentSpace.GetName()),
+		)
+		// this is a grandchild
+		grandChildSpace1 := spacetest.NewSpace(test.HostOperatorNs, "smith-grandChild1",
+			spacetest.WithTierName("advanced"),
+			spacetest.WithSpecTargetCluster(test.MemberClusterName),
+			spacetest.WithSpecParentSpace(childSpace1.GetName()),
+		)
+		spaces := map[string]*toolchainv1alpha1.Space{
+			parentSpace.Name: parentSpace,
+			childSpace1.Name: childSpace1, childSpace2.Name: childSpace2,
+			grandChildSpace1.Name: grandChildSpace1,
+		}
+		// the defaultGetSpaceFunc returns the space based on the given name
+		defaultGetSpaceFunc := func(spaceName string) (*toolchainv1alpha1.Space, error) {
+			if space, found := spaces[spaceName]; found {
+				return space, nil
+			}
+			return nil, fmt.Errorf("space not found: %s", spaceName)
+		}
+
+		// listParentSpaceBindingFunc returns the spacebinding for the parentSpace
+		listParentSpaceBindingFunc := func(spaceName string) ([]toolchainv1alpha1.SpaceBinding, error) {
```

Comment:
This function looks wrong TBH. You don't really test your code properly with that. I would just create a singe function which would do that switch with all the cases (it returns SBs for all the spaces depending on the `spaceName`). And I would use that single function in all the tests. Basically you could just merge all the custom functions you use below to a single one. Something like:

```
func(spaceName string) ([]toolchainv1alpha1.SpaceBinding, error) {
	switch spaceName {
	case parentSpace.GetName():
		return []toolchainv1alpha1.SpaceBinding{
			*spacebinding.NewSpaceBinding(mur, parentSpace, parentSpace.Name, spacebinding.WithRole("admin")),
		}, nil
	case grandChildSpace1.GetName():
		return []toolchainv1alpha1.SpaceBinding{
			*spacebinding.NewSpaceBinding(mur, grandChildSpace1, grandChildSpace1.Name, spacebinding.WithRole("viewer")),
		}, nil
	...
	default:
		return nil, nil
	}
},
```
In the runtime code you use the same listSB function. Which returns SBs for the provided space. In the test, however, your mock function returns only the expected SBs.
For example, in the first test the mock function will always return the parent SB no matter what space the runtime code asked the SBs for. It can hide bugs in the code.

But I like the new tests much better overall! Good job!

Reply:
If we always return a spacebinding we will not test the case in which the role is inherited from one of the parent spaces right ?

Reply:
In the runtime you provide the same function which returns the actual SBs for the provided space, right?
Shouldn't we do the same in the test? You specify the same single space hierarchy for all the tests. You already know what SBs should be return for a specific space. And this function, in the runtime code, doesn't know anything about the inherited roles. It just returns SBs CRs for the space. It's the job of the lister to look up for SBs in all the parent spaces, so the inherited SBs are included too.
Shouldn't we do the same in the test?
Or I'm missing something?

Reply:
> In the runtime you provide the same function which returns the actual SBs for the provided space, right?

Not always. I think this is more close to what happens at runtime because in the case of a subSpace (let's say created with `SpaceRequest`) there is no `SpaceBinding` resource directly linked with it, by default it inherits the parent spacebindings only. So this call here will return empty list https://github.com/codeready-toolchain/toolchain-common/pull/329/files#diff-c83558581efe8e9049f110acd8dc83b158f52beeab1b6da8ee1299ab9ccda983R24
when called with the subSpace.
But the function will go on and search for the spaceBindings associated with the parentSpace and will return them here: https://github.com/codeready-toolchain/toolchain-common/pull/329/files#diff-c83558581efe8e9049f110acd8dc83b158f52beeab1b6da8ee1299ab9ccda983R39-R45

> And this function, in the runtime code, doesn't know anything about the inherited roles. It just returns SBs CRs for the space. It's the job of the lister to look up for SBs in all the parent spaces, so the inherited SBs are included too.

The lister also merges the roles based on the MUR, the child one overrides the parent ones. So if I well understood your proposal - to  always return a spacebinding resource for each space, we are already covering this case here https://github.com/codeready-toolchain/toolchain-common/pull/329/files#diff-8467fbd206b9f86061c7253992813af5478d514bd8061a26920f13fc5042ffb0R142-R168

The other test cases are also covering the cases in which this call https://github.com/codeready-toolchain/toolchain-common/pull/329/files#diff-c83558581efe8e9049f110acd8dc83b158f52beeab1b6da8ee1299ab9ccda983R24 returns an empty list, but subsequent (parent spaces) or previous calls (grandchild) return a spacebinding.

Reply:
Following is a list of the hierarchies we cover with this code:

1) parentSpace has it's own spacebinding
```
johnny
|__ parentSpace -> parentSpaceBinding role=admin
```
**result:** list for `parentSpace` expects to return `admin` 

2) spacebinding for childSpace1 is inherited from parentSpace
```
johnny
|__ parentSpace        ---> parentSpaceBinding role=admin
      |__ childSpace1  ---> no specific spacebinding, inherits the one above role=admin
```
**result:** list for `childSpace1` expects to return `admin` 

3) override parentSpace-binding in childSpace2
```
johnny
|__ parentSpace       ---> parentSpaceBinding role=admin
      |__ childSpace1 ---> childSpace2Binding role=viewer
```
**result:** list for `childSpace1` expects to return `viewer`

4) no spacebinding on child1 and grandChildSpace1, spacebinding for grandChildSpace1 is inherited from parentSpace
```
johnny
|__ parentSpace                   ---> parentSpaceBinding role=admin
      |__ childSpace1             ---> no specific spacebinding, inherits the one above role=admin
            |__ grandChildSpace1  ---> no specific spacebinding, inherits the one from parentSpace role=admin because of the recursive searches up for parentSpaces  
```
**result:** list for `grandChildSpace1` expects to return `admin` 

5) no spacebinding on child1 but we do have one on grandChildSpace1 which overrides parentSpace-binding
```
johnny
|__ parentSpace                  ---> parentSpaceBinding role=admin
      |__ childSpace1            ---> no specific spacebinding, inherits the one above role=admin
            |__ grandChildSpace1 ---> grandChildSpace1Binding role=viewer   
```
**result:** list for `grandChildSpace1` expects to return `viewer`

6) spacebinding on grandChildSpace1 overrides child1-binding and parentSpace-binding
```
johnny
|__ parentSpace                  ---> parentSpaceBinding role=admin
      |__ childSpace1            ---> childSpace1Binding role=maintainer
            |__ grandChildSpace1 ---> grandChildSpace1Binding role=viewer   
```
**result:** list for `grandChildSpace1` expects to return `viewer`

NOTE: I can create 2 other use-cases based on this last test case , to list for `parentSpace` and `grandChildSpace1` but not sure how much additional value this will bring since the mock will always return only that specific spacebinding associated with the space and the merge function will override the parent ones. But I can add those.

Now that I look at this, what is still missing IMO, is some test case with more MURs, so that the listing returns more than one spacebinding for different users, maybe with different roles.

For example:
```
johnny
|__ parentSpace                  ---> parentSpaceBinding role=admin
      |__ childSpace1            ---> no specific spacebinding, inherits the one above role=admin
            |__ grandChildSpace1 ---> no specific spacebinding, inherits the one above role=admin
batman
|__ parentSpace                  ---> no access
      |__ childSpace1            ---> childSpace1Binding role=maintainer
            |__ grandChildSpace1 ---> no specific spacebinding, inherits the one above role=maintainer
robin
|__ parentSpace                  ---> no access
      |__ childSpace1            ---> no access
            |__ grandChildSpace1 ---> grandChildSpace1Binding role=viewer   
```
**results:**
list for _parentSpace_ should return _johnny role=admin_
list for _childSpace1_ should return _johnny role=admin_ , _batman role=maintainer_
list for _grandChildSpace1_ should return _johnny role=admin_ , _batman role=maintainer_ , _robin role=viewer_

sorry for the long comment! 

WDYT?

Reply:
@alexeykazakov as discussed yesterday, I've updated the test which now builds the hierarchy upfront , I've also added a doc file with the hierarchy diagram (note needs mermaid plugin locally, but github should render that by default)

please see: https://github.com/codeready-toolchain/toolchain-common/pull/329/commits/9e0ee2b48d1177745f66bf43a8feff148336fcb5